warehouse/query-core/src/main/java/datawave/query/util/sortedset/BufferedFileBackedSortedSet.java

```diff
@@ -156,56 +158,78 @@ public boolean containsAll(Collection<?> c) {
-                int iterationsWithExtra = numSets / setsPerCompaction;
+            // create a copy of the set list (sorting below)
+            List<SortedSet<E>> sets = new ArrayList<>(set.getSets());
+            
+            // calculate the number of sets to compact per round
+            int numSets = sets.size();
+            int excessSets = numSets - (maxFiles / 2); // those over 50% of maxFiles
+            int filesToCompact = excessSets + rounds; // Add in a few extra to account for the compacted sets being added back in
+            int setsPerCompaction = (filesToCompact + rounds - 1) / rounds; // divide into sets to compact, rounding up
+            
+            // sort the sets by size (compact up smaller sets first)
+            sets.sort(Comparator.comparing((SortedSet<E> s) -> s.size()));
+
+            // newSet will be the final multiset
+            MultiSetBackedSortedSet<E> newSet = new MultiSetBackedSortedSet<>();
+            
+            for (int r = 0; r < rounds; r++) {
                 
-                MultiSetBackedSortedSet<E> newSet = new MultiSetBackedSortedSet<>();
+                // create a set for those sets to be compacted into one file
                 MultiSetBackedSortedSet<E> setToCompact = new MultiSetBackedSortedSet<>();
-                
-                int iteration = 0;
-                byte subtraction = (byte) (iteration < iterationsWithExtra ? 0 : 1);
-                for (int i = 0; i < numSets; i++) {
+                for (int i = 0; i < setsPerCompaction; i++) {
                     setToCompact.addSet(sets.get(i));
-                    if (setToCompact.getSets().size() == (setsPerCompaction - subtraction)) {
-                        newSet.addSet(compact(setToCompact));
-                        setToCompact.clear();
-                        iteration++;
-                        subtraction = (byte) (iteration < iterationsWithExtra ? 0 : 1);
-                    }
                 }
-                if (!setToCompact.isEmpty()) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("Starting compaction for " + setToCompact);
-                    }
-                    long start = System.currentTimeMillis();
-                    FileSortedSet<E> compaction = compact(setToCompact);
-                    if (log.isDebugEnabled()) {
-                        long delta = System.currentTimeMillis() - start;
-                        log.debug("Compacted " + setToCompact + " -> " + compaction + " in " + delta + "ms");
-                    }
-                    newSet.addSet(compaction);
-                    setToCompact.clear();
+                
+                // remove the sets being compacted from our list
+                // creating new array list to allow previous list to be GC'ed
+                sets = new ArrayList<>(sets.subList(setsPerCompaction, sets.size()));
```

Comment:
You're already working with a copy (made on line 182). Shouldn't you just be able to do
```java
sets = sets.subList(setsPerCompaction, sets.size());
```
and avoid the extra object creation?

Reply:
Or on line 201 why not just call remove until you have the number you want:
```
setToCompact.addSet(sets.remove(0));
```

Reply:
Since part of the goal was to avoid some memory issues we were having, I was attempting to ensure that the old sets could be garbage collected as we compacted things away.  However now that I think about it, the global set is still holding on to these references so it does not matter.  I actually was using subList originally.  I will revert back to that.

Reply:
I kinda like the remove idea.  Perhaps I will reverse the order of the list to allow a more efficient remove or changed to a LinkedList.....